Wait for the network to be online before starting tinc. (1.1)

fangfufu opened the pull request:
I found it super annoying that tinc starts before any of my other network interfaces come online. I made this change locally, so tinc starts up after the network comes online. OpenVPN has similar behaviour. I think it might be useful for others.

Note that the behaviour of tinc not waiting until the network comes online is mentioned in issue https://github.com/gsliepen/tinc/issues/133.

There is also a similar pull request for tinc 1.0. (pull request https://github.com/gsliepen/tinc/pull/325)

hg commented:
By the way, I also find these changes useful:

```ini
[Service]
User = tinc
Group = tinc
AmbientCapabilities = CAP_NET_ADMIN CAP_NET_BIND_SERVICE
```

tinc does not really require root in most circumstances. `CAP_NET_BIND_SERVICE` allows it to bind to 655, and `CAP_NET_ADMIN` to configure network interfaces (including processes started through tincd scripts). The second one is a very powerful capability, but it's still better than root.

fangfufu commented:
By default tinc doesn't create its own user. At least not under Debian. If we want this in, then we should really create new user and new group as a part of the installation.

hg commented:
Well, there's `DynamicUser` when systemd is recent enough.

http://0pointer.net/blog/dynamic-users-with-systemd.html

It's probably worth getting in before 1.1 is finally released, which is the best time for breaking changes. tinc is written in C, after all, we need all the hardening we can get.

hg commented:
I also suggest looking at

    $ systemd-analyze security tinc@

if your systemd is not very ancient. There are a ton of isolation options (many of those are used by various container engines). I've been doing this with work-related stuff for many years.

fangfufu commented:
``DynamicUser`` requires at least Debian Buster, which is now the OldStable. So I think we should put it in. 
https://packages.debian.org/search?suite=all&searchon=names&keywords=systemd

fangfufu commented:
> 
> 
> By the way, I also find these changes useful:
> 
> ```ini
> [Service]
> User = tinc
> Group = tinc
> AmbientCapabilities = CAP_NET_ADMIN CAP_NET_BIND_SERVICE
> ```
> 
> tinc does not really require root in most circumstances. `CAP_NET_BIND_SERVICE` allows it to bind to 655, and `CAP_NET_ADMIN` to configure network interfaces (including processes started through tincd scripts). The second one is a very powerful capability, but it's still better than root.

Added

hg commented:
This probably won't work as it's currently written, because by default tinc config directory belongs to root, and blocks access to keys to users others than root.

This kludge should fix it:

`ExecStartPre=+/usr/bin/chown -R tinc:tinc @sysconfdir@/tinc/%i`

but the best solution would be changing installation (and upgrade) scripts and doing it there.

fangfufu commented:
> but the best solution would be changing installation (and upgrade) scripts and doing it there.

Personally I think this should be split into two pull requests - one to add my original proposed changes, the second one does the ``DynamicUser`` and the directory/file ownership changes.

fangfufu commented:
> 
> 
> By the way, I also find these changes useful:
> 
> ```ini
> [Service]
> User = tinc
> Group = tinc
> AmbientCapabilities = CAP_NET_ADMIN CAP_NET_BIND_SERVICE
> ```
> 
> tinc does not really require root in most circumstances. `CAP_NET_BIND_SERVICE` allows it to bind to 655, and `CAP_NET_ADMIN` to configure network interfaces (including processes started through tincd scripts). The second one is a very powerful capability, but it's still better than root.

I think I am not going to add these in this pull request, as I cannot get it them working locally for my tinc 1.0. (Yes, I know this is pull request is for tinc 1.1) We need to change the installation script, as you stated in https://github.com/gsliepen/tinc/pull/326#issuecomment-900203979.

The results of my attempts are documented in https://github.com/gsliepen/tinc/pull/325#issuecomment-900207241 and https://github.com/gsliepen/tinc/pull/325#issuecomment-900220257. 

We should have a separate pull request which changes the installation script. I am not sure what to do right now, so I am not adding them in this pull request.

gsliepen commented:
Ah, the dreaded network-online target. For some people, the network isn't online until the VPN is up and running, and they could've configured their systemd setup so! Also, tinc should not have to depend on the network being online, as it's supposed to handle network outages to begin with, so the network not being available the first few seconds after it starts should be fine.

> I found it super annoying that tinc starts before any of my other network interfaces come online.

Why is this annoying exactly? Just in case there is a better solution that addresses the issues you are getting because of this.

fangfufu commented:
>  For some people, the network isn't online until the VPN is up and running, and they could've configured their systemd setup so!
Now you mentioned it, the network-online target does sound a bit ambiguous. However surely the physical network should be up and running before VPN starts up? 

> Why is this annoying exactly? Just in case there is a better solution that addresses the issues you are getting because of this.

I basically have a bridge which bridge my different layer-2 VPN interfaces together. There is no easy way to add tinc to the bridge if tinc starts before ``/etc/network/interfaces`` gets configured.

splitice commented:
Support for systemd notify (on tinc ready) would also be nice (seperate). That's what I first thought when I read the title.

gsliepen commented:
How about adding tinc's interface to the bridge in the [host-up](https://www.tinc-vpn.org/documentation-1.1/Scripts.html) script instead? But yes, in that case a dependency on whatever brings the bridge up would be nice.

gsliepen commented:
> Support for systemd notify (on tinc ready) would also be nice (seperate).

This could be put in the tinc-up script, or also in a host-up script if you want the notification to happen only when tinc has actually connected to another node.

splitice commented:
Interesting idea, it could work. There can be pid issues with approaches like that but I think uid 0 has a pass.

gsliepen commented:
> tinc does not really require root in most circumstances. CAP_NET_BIND_SERVICE allows it to bind to 655, and CAP_NET_ADMIN to configure network interfaces (including processes started through tincd scripts). The second one is a very powerful capability, but it's still better than root.

It really depends on what people put in their scripts. That said, safer defaults for tinc 1.1 might be a good idea. I wouldn't use this for tinc 1.0, as that will just upset users that do need the extra capabilities and that expect it to be stable.

hg commented:
> It really depends on what people put in their scripts.

I've been using it everywhere. If tincd requires root access to configure something on one machine, I throw it into a separate script (owned by root), and give the `tinc` user the appropriate sudo privileges:

    tinc ALL = (ALL) NOPASSWD: /etc/tinc/net/tinc-up

fangfufu commented:
> > There is no easy way to add tinc to the bridge if tinc starts before /etc/network/interfaces gets configured.
> 
> How about adding tinc's interface to the bridge in the [host-up](https://www.tinc-vpn.org/documentation-1.1/Scripts.html) script instead? But yes, in that case a dependency on whatever brings the bridge up would be nice.

That's a good idea. Maybe I should just close this PR. One last question thing though, would it be a good idea if we align our behaviour with OpenVPN?

gsliepen commented on systemd/tinc@.service.in:
```diff
@@ -4,6 +4,9 @@ Documentation=info:tinc
 Documentation=man:tinc(8) man:tinc.conf(5)
 Documentation=http://tinc-vpn.org/docs/
 PartOf=tinc.service
+Before=systemd-user-sessions.service
```
OpenVPN doesn't have this requirement. Why do we need it?

fangfufu replied:
I copied from Debian Bullseye's ``/lib/systemd/system/openvpn@.service``.
```
fangfufu@smithsonian:~$ cat /lib/systemd/system/openvpn@.service
[Unit]
Description=OpenVPN connection to %i
PartOf=openvpn.service
Before=systemd-user-sessions.service
After=network-online.target
Wants=network-online.target
Documentation=man:openvpn(8)
Documentation=https://community.openvpn.net/openvpn/wiki/Openvpn24ManPage
Documentation=https://community.openvpn.net/openvpn/wiki/HOWTO

[Service]
Type=notify
PrivateTmp=true
WorkingDirectory=/etc/openvpn
ExecStart=/usr/sbin/openvpn --daemon ovpn-%i --status /run/openvpn/%i.status 10 --cd /etc/openvpn --config /etc/openvpn/%i.conf --writepid /run/openvpn/%i.pid
PIDFile=/run/openvpn/%i.pid
KillMode=process
CapabilityBoundingSet=CAP_IPC_LOCK CAP_NET_ADMIN CAP_NET_BIND_SERVICE CAP_NET_RAW CAP_SETGID CAP_SETUID CAP_SYS_CHROOT CAP_DAC_OVERRIDE CAP_AUDIT_WRITE
LimitNPROC=100
DeviceAllow=/dev/null rw
DeviceAllow=/dev/net/tun rw
ProtectSystem=true
ProtectHome=true
RestartSec=5s
Restart=on-failure

[Install]
WantedBy=multi-user.target
```

Mic92 replied:
I don't see why tinc needs this:

```
       systemd-user-sessions.service is a service that controls user logins
       through pam_nologin(8). After basic system initialization is complete,
       it removes /run/nologin, thus permitting logins. Before system
       shutdown, it creates /run/nologin, thus prohibiting further logins.
```

Tinc does not require logins or pam of any sorts.

gsliepen commented:
I see that OpenVPN's service file has some useful things, like `DeviceAllow=/dev/net/tun`, but it's missing a `Wants+After=modprobe@tun.service` then.

I also wonder what the best way is to run tinc as non-root under systemd. I guess systemd.networkd could be used to create and configure a tun or tap device that can be accessed by a non-root user, and then let tincd use that one, and not have up/down scripts. But I don't see how this could work together with `DynamicUser`.

fangfufu commented:
I kind of like having up/down scripts. This allows the end users to do more advanced configurations, like adding the tun/tap device to a bridge.

hg commented:
>I guess systemd.networkd could be used to create and configure a tun or tap device that can be accessed by a non-root user

That's what I've been using, although the device is still being created by tinc (using `CAP_NET_ADMIN`).

    $ cat /etc/systemd/network/tinc@tinc0.network

```ini
[Match]
Name = tinc0

[Network]
Address = 10.99.0.1/16
LinkLocalAddressing = no

[Route]
Destination = 10.10.19.0/24
Destination = 192.168.50.0/24
```

`DynamicUser` is not really required, we could use a regular user as has been done for the past 40 years. Every other isolation option that it enables (like `ProtectSystem`) can be (and should be) enabled separately. I am using this, which is not the strictest possible policy and can be improved further (but prevents access to almost the whole system, including all user files, and should prevent RCEs):

```ini
UMask                   = 0077
LockPersonality         = yes
NoNewPrivileges         = yes
PrivateMounts           = yes
PrivateTmp              = yes
ProtectClock            = yes
ProtectControlGroups    = yes
ProtectHome             = yes
ProtectHostname         = yes
ProtectKernelLogs       = yes
ProtectKernelModules    = yes
ProtectKernelTunables   = yes
ProtectSystem           = strict
ReadWritePaths          = /etc/tinc
MemoryDenyWriteExecute  = yes
DevicePolicy            = closed
DeviceAllow             = /dev/net/tun
RestrictAddressFamilies = AF_UNIX AF_INET AF_INET6
RestrictNamespaces      = yes
RestrictRealtime        = yes
RestrictSUIDSGID        = yes
SystemCallArchitectures = native
ProtectProc             = invisible
CapabilityBoundingSet   = CAP_NET_ADMIN CAP_NET_BIND_SERVICE
SystemCallFilter        = ~@clock @module @mount @reboot @swap @privileged @cpu-emulation @obsolete
```

Because you have full control over the code base and know every system call that tinc could possibly make unless overtaken by script kiddies, the seccomp-bpf system call filter in `SystemCallFilter` can be trimmed down to absolute necessities instead of using broad groups like I'm doing above.

---

>like adding the tun/tap device to a bridge.

This can be done through systemd-networkd. Some pointers [here](https://wiki.archlinux.org/title/Systemd-networkd#Bridge_interface), and also take a look at `man systemd.network` → `/` `Example 4.`.

SilverBut commented:
Maybe not the best way to implement this feature, as `network-online.target` might mean "tinc is up" for some users, and it's changing behaviors for current service.

If users really want `tinc` start triggered by `network-online`, what about using [drop-in files](https://wiki.archlinux.org/title/systemd#Drop-in_files), and add this into the manual?

Mic92 commented on systemd/tinc@.service.in:
```diff
@@ -4,6 +4,9 @@ Documentation=info:tinc
 Documentation=http://tinc-vpn.org/docs/
 PartOf=tinc.service
+Before=systemd-user-sessions.service
+After=network-online.target
```
I am strongly against depending on `network-online.target` by default as this target might be never reached if either networkd or network-manager do not get a DHCP address. Tinc would be still useful if there are other nodes in the same ethernet segment, which it can connect to through broadcast discovery.
If someone wants to create devices before starting tinc a better way is to add dependencies on devices units.
Tinc is decentral and does not depend on a default gateway hence `network-online.target` is not needed.

Mic92 commented:
> > For some people, the network isn't online until the VPN is up and running, and they could've configured their systemd setup so!
> > Now you mentioned it, the network-online target does sound a bit ambiguous. However surely the physical network should be up and running before VPN starts up?
> 
> > Why is this annoying exactly? Just in case there is a better solution that addresses the issues you are getting because of this.
> 
> I basically have a bridge which bridge my different layer-2 VPN interfaces together. There is no easy way to add tinc to the bridge if tinc starts before `/etc/network/interfaces` gets configured.

Use device units for this.

fangfufu commented:
@Mic92 , is there a good example for the unit file? 

Also, closing this pull request, because I am convinced that this is probably not a great idea.

Mic92 commented:
@fangfufu 

You can get device units created like this:

```console
$ systemctl list-units | grep sys-subsystem-net-devices-
  sys-subsystem-net-devices-enp0s31f6.device                                             loaded active plugged   Ethernet Connection (10) I219-V
  sys-subsystem-net-devices-tinc.retiolum.device                                         loaded active plugged   /sys/subsystem/net/devices/tinc.retiolum
  sys-subsystem-net-devices-wlan0.device                                                 loaded active plugged   Comet Lake PCH-LP CNVi WiFi (Wi-Fi 6 AX201 160MHz)
```

You unit should than contain the wanted device:

```
After = sys-subsystem-net-devices-tinc.retiolum.device
Requires = sys-subsystem-net-devices-tinc.retiolum.device 
```
